ehrql/tables/beta/tpp.py

```diff
@@ -586,129 +653,237 @@ class occupation_on_covid_vaccine_record(EventFrame):
 
 @table
 class opa(EventFrame):
+    """
+    Outpatient appointments data is provided via the NHS Secondary Uses Service.
+
+    This table gives core details of outpatient appointments.
+    """
+
     opa_ident = Series(
         int,
         constraints=[Constraint.NotNull()],
-        description="TODO",
+        description="Unique identifier for the appointment used across the OPA tables.",
     )
     appointment_date = Series(
         datetime.date,
-        description="TODO",
+        description="The date of an appointment.",
     )
     attendance_status = Series(
         str,
-        description="TODO",
+        description=(
+            "Indicates whether or not an appointment for a care contact took place. "
+            "If the appointment did not take place it also indicates whether or not advanced warning was given. "
+            "Refer to the [NHS Data Model and Dictionary entry for attendance status]"
+            "(https://www.datadictionary.nhs.uk/data_elements/attendance_status.html).\n"
+            "Permitted codes:\n"
+            "* `5`: Attended on time or, if late, before the relevant care professional was ready to see the patient\n"
+            "* `6`: Arrived late, after the relevant care professional was ready to see the patient, but was seen\n"
+            "* `7`: Patient arrived late and could not be seen\n"
+            "* `2`: Appointment cancelled by, or on behalf of, the patient\n"
+            "* `3`: Did not attend - no advance warning given\n"
+            "* `4`: Appointment cancelled or postponed by the health care provider"
+        ),
+        # This non-ascending order follows the NHS Data Model and Dictionary.
+        constraints=[Constraint.Categorical(["5", "6", "7", "2", "3", "4"])],
```

Comment:
question: For this and the other categorical constraints, I have taken the values from the NHS Data Model and Dictionary. I don't have the access to investigate whether TPP adheres to these strictly or does something else; for example, zero pad the values.

All of these categorical columns are varchars with length 2).

Reply:
There are `0` and `NULL` values in the DB.

I'd suggest leaving the mapping from codes to names for another day. I'm sure there are other examples of tables in this namespace that are unmapped. There's a bigger issue that relates to making tables less raw and raw tables more raw.